[Metrics builder] Move resource creation to the generated code

dmitryax opened the pull request:
This change updates metrics builder to fully control creation of the `pdata.Metrics` object, scraper author don't need to create and pass it to `MetricsBuiler.Emit` function anymore.

This change also moves definition of resource attributes to metadata.yaml which is used to generate attribute helper that can be used in the scraper.

Closes: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/8358

Also this is the first PR towards https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/7301.

The following items will be resolved as follow-up PRs:
- Add sem ver support in metadata.yaml
- Define resource schema in metadata.yaml and automatically set it up instead of using `wrapBySchemaURLSetterConsumer`

dmitryax commented:
@djaglowski @codeboten please let me know what do you think about this approach to https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/8358

mstumpfx commented on CHANGELOG.md:
```diff
@@ -2,6 +2,11 @@
 
+### 💡 Enhancements 💡
+
+- `cmd/mdatagen`:disk_scraper_windows.go Add resource attributes definition to metadata.yaml and move `pdata.Metrics` craetion to the 
```
```suggestion
- `cmd/mdatagen`: Add resource attributes definition to metadata.yaml and move `pdata.Metrics` creation to the 
```

dmitryax replied:
Good catch. Thanks :)

djaglowski replied:
These changes appear not to have actually been applied.

dmitryax replied:
Oh I overridden them locally. Let me recover it

dmitryax replied:
fixed

codeboten commented on receiver/couchdbreceiver/documentation.md:
```diff
@@ -30,7 +30,6 @@ metrics:
 
 | Name | Description |
 | ---- | ----------- |
-| couchdb.node.name | The name of the node. |
```
It looks like the documentation generator needs updating here.

dmitryax replied:
Yes, this is because documentation generator doesn't get resource attributes yet. I have in the list of follow-up PRs, but I can add it to this one if you think it's needed

dmitryax replied:
Added docs generation in this PR

codeboten commented on receiver/rabbitmqreceiver/documentation.md:
```diff
@@ -29,6 +29,3 @@ metrics:
 | message.state | The state of messages in a queue. |
-| rabbitmq.node.name | The name of the RabbitMQ node. |
-| rabbitmq.queue.name | The name of the RabbitMQ queue. |
-| rabbitmq.vhost.name | The name of the RabbitMQ vHost. |
```
It looks like the documentation generator needs updating here.

dmitryax replied:
same here

codeboten commented on receiver/zookeeperreceiver/documentation.md:
```diff
@@ -38,6 +38,4 @@ metrics:
 | Name | Description |
 | ---- | ----------- |
 | direction | State of a packet based on io direction. |
-| server.state | State of the Zookeeper server (leader, standalone or follower). |
```
It looks like the documentation generator needs updating here.

dmitryax replied:
done

codeboten reviewed (commented):
Overall I think this looks really good. Just a couple of things that need to be fixed w/ the generated documentation.

djaglowski commented on receiver/couchdbreceiver/scraper.go:
```diff
@@ -92,10 +79,5 @@ func (c *couchdbScraper) appendMetrics(stats map[string]interface{}, rms pdata.R
-	}
-
-	return errors.Combine()
+	return c.mb.Emit(metadata.WithCouchdbNodeNameResourceAttribute(c.config.Endpoint)), errors.Combine()
```
nit: The function names are always going to be very verbose if at a minimum we have `WithFooResourceAttribute`.  We could consider shortening the function name. I think it still reads fine to the scraper author. 
```suggestion
	return c.mb.Emit(metadata.WithCouchdbNodeName(c.config.Endpoint)), errors.Combine()
```

dmitryax replied:
I agree. Updated the code to the short version

djaglowski commented on receiver/rabbitmqreceiver/internal/metadata/generated_metrics_v2.go:
```diff
@@ -392,16 +396,73 @@ func NewMetricsBuilder(settings MetricsSettings, options ...metricBuilderOption)
+// recording another set of metrics. This function will be responsible for applying all the transformations required to
+// produce metric representation defined in metadata and user settings, e.g. delta or cumulative.
+func (mb *MetricsBuilder) Emit(options ...ResourceOption) pdata.Metrics {
+	mb.EmitForResource(options...)
```
I really like this pattern overall but it isn't clear to me if it creates an extra resource when used like this:
```go
for _, resource := range resources {
    ...
    mb.EmitForResource(metadata.WithFooResourceAttribute(resource.foo))
}
mb.Emit()
```

I think this is a common pattern for collecting, but it looks like maybe it will create an empty resource at the end. Possibly I am missing how this is avoided but want to make sure.

dmitryax replied:
No empty resources created in this case. Metric builder always trims empty resources

dmitryax commented on receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go:
```diff
@@ -118,10 +108,11 @@ func (s *scraper) scrape(_ context.Context) (pdata.Metrics, error) {
 		}
-		s.mb.Emit(metrics)
+
+		s.mb.EmitForResource(md.resourceOptions()...)
```
Makes sense. Created an issue https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/8815

dmitryax committed:
[Metrics builder] Move resource creation to the generated code

This change updates metrics builder to fully control creation of the metrics object. Scraper author don't need to create pdata.Metrics object and pass it to MetricsBuiler.Emit function.

This change also moves definition of resource attributes to metadata.yaml which is used to generate attribute helper that can be used in the scraper.
313873

djaglowski reviewed (approved):
LGTM. This is a really nice solution to two problems.
